asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_heat_pump_keywords.py

```diff
@@ -0,0 +1,227 @@
+        .replace("gshps", "gshp")
+        .replace("gshp", "ground source heat pump")
+        .replace("hps", "hp")
+        .replace("hp", "heat pump")
+        .replace("ufh", "under floor heating")
+    )
+)
+bh_data["text"] = (
+    bh_data["text"]
+    .astype(str)
+    .apply(
+        lambda x: x.lower()
+        .replace("ashps", "ashp")
+        .replace("ashp", "air source heat pump")
+        .replace("gshps", "gshp")
+        .replace("gshp", "ground source heat pump")
+        .replace("hps", "hp")
+        .replace("hp", "heat pump")
+        .replace("ufh", "under floor heating")
+    )
+)
+
+# %%
+mse_data["category"].unique()
+
+# %%
+keywords = "heat_pump_keywords"
+
+# %% [markdown]
+# ## MSE titles mentioning "heat pump"
+
+# %%
+hp_data_mse = mse_data[
+    mse_data["title"].str.contains("|".join(keyword_dictionary[keywords]), case=False)
+    | mse_data["text"].str.contains("|".join(keyword_dictionary[keywords]), case=False)
+]
+
+# %%
+docs = hp_data_mse[hp_data_mse["is_original_post"] == 1].drop_duplicates("title")[
+    "title"
+]
+
+# %%
+len(docs)
+
+# %%
+
+umap_model = UMAP(
+    n_neighbors=15, n_components=5, min_dist=0.0, metric="cosine", random_state=42
+)
+topic_model = BERTopic(umap_model=umap_model)
+topics, probs = topic_model.fit_transform(docs)
+
+
+# %%
+topics, topics_info, doc_info = get_outputs_from_topic_model(topic_model, docs)
+
+# %%
+topics_info.head()
+
+# %%
+doc_info.head()
+
+# %%
+create_bar_plot_most_common_topics(topics_info=topics_info, top_n_topics=16)
+
+# %%
+topic_model.visualize_topics()
+
+# %%
+topic_model.visualize_barchart(
+    topics=range(15), title="Topic word scores - Heat Pump Mentions in MSE"
+)
+
+# %%
+topic_model.visualize_documents(
+    docs=list(docs),
+    width=1200,
+    height=750,
+    topics=range(15),
+    title="Top 15 topics in Heat Pump mentions in MSE",
+)
+
+# %%
+topic_model.visualize_hierarchy()
+
+# %%
+topic_model.visualize_term_rank()
+
+
+# %%
+
+
+# %% [markdown]
+# ## BuildHub titles mentioning "heat pump"
+
+# %%
+hp_data_bh = bh_data[
+    bh_data["title"].str.contains("|".join(keyword_dictionary[keywords]), case=False)
+    | bh_data["text"].str.contains("|".join(keyword_dictionary[keywords]), case=False)
+]
+
+# %%
+docs = hp_data_bh[hp_data_bh["is_original_post"] == 1].drop_duplicates("title")["title"]
+
+# %%
+len(docs)
+
+# %%
+umap_model = UMAP(
+    n_neighbors=15, n_components=5, min_dist=0.0, metric="cosine", random_state=42
+)
+# vectorizer_model = CountVectorizer(stop_words="english")
+# representation_model = KeyBERTInspired()
+
+topic_model = BERTopic(umap_model=umap_model)
+topics, probs = topic_model.fit_transform(docs)
+
+
+# %%
+topics, topics_info, doc_info = get_outputs_from_topic_model(topic_model, docs)
+
+# %%
+topics_info.head()
+
+# %%
+doc_info.head()
+
+# %%
+create_bar_plot_most_common_topics(topics_info=topics_info, top_n_topics=16)
+
+# %%
+topic_model.visualize_topics()
```

Comment:
For all these, it would be great to be able to save the figures in a png so we can have a look at the outputs
e.g. 
`  fig = topic_model.visualize_topics()
    fig.write_image(figure_file_path + "topic_visualisation.png")`. I know this was initially done as a notebook, so I guess I can convert it back to that to view the images.

Reply:
yep, at this point this is not the final script, so I didn't want to spend time writing code to save the figures. At the top of the file, in the docstring, I explain how to open these as notebooks in order to visualise the results.